### PR TITLE
Audiobook fixes

### PR DESCRIPTION
**Pull request by dimitry-ishenko**
## Description

These patches fix audio book resume functionality.

## Motivation and Context

See issue #15216 for details.

## How Has This Been Tested?

Played an .mka file referenced in issue #15216.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by dimitry-ishenko** [comment]
These patches fix most of the problems mentioned in #15216. Specifically:

- There is now single bookmark per audio book
- Opening context menu for any chapter within the audio book will have "Resume audiobook" menu item
- The audio book will resume playing from the right chapter
- Strange screen no longer shows up

One thing that's not working is the audio book starts playing from the beginning of the chapter even if it was stopped in the middle. If you look in the last patch, I had to comment out lines 700 and -701. If I uncomment them, for some reason the playback starts from the very beginning (even though the correct chapter is shown on the screen).

Could someone point me in the right direction what to do here?

**Comment by dimitry-ishenko** [comment]
OK, I found the problem:

1. PR #14649 by @angelblue05 is partially responsible for this, as audio book chapters rely on `m_lStartOffset` and `m_lEndOffset` and setting `m_lStartOffset` to 0 causes playback to start from the beginning.
2. This code:
https://github.com/xbmc/xbmc/blob/cb6e844ac61771fb0328fdd06d62fbb0046a02aa/xbmc/cores/paplayer/PAPlayer.cpp#L260-L264
is the second problematic part. 50ms sleep is not long enough on my laptop and neither is 100ms. If I set it to 150ms, I hear a small blip from the beginning of the chapter before it seeks to the bookmark. I think method needs to be reworked, as it will never be reliable.

**Comment by kib** [comment]
I think you are wrong on (1).
That PR allows a skinner to create a button which manually sets StartOffSet to 0. (In case you want to force a 'start from beginning') If you do not call it like that it does not set it to 0, or do you mean that it always becomes zero even when not called like it.

**Comment by kib** [comment]
Never mind, I see what you meant.

**Comment by angelblue05** [comment]
Can you point to where StartPercent is set?

**Comment by dimitry-ishenko** [comment]
> Can you point to where StartPercent is set in your scenario?

Over here:
https://github.com/xbmc/xbmc/blob/cb6e844ac61771fb0328fdd06d62fbb0046a02aa/xbmc/music/windows/GUIWindowMusicBase.cpp#L690-L709

Specifically, lines 704-706

**Comment by dimitry-ishenko** [comment]
@angelblue05 I've read through #14649 and couldn't fully understand the problem it solves. Could you maybe give me an example of an addon where it is used?

Another thing I've noticed:
The `StartPercent` property seems to be treated differently. For audio books, if eg `StartPercent == 10%` that means play the current list item from `m_lStartOffset` + 10% **of the current item length**.
I believe in your case that means start playing from beginning of the entire list + 10%, whichever list item that might be. Is my understanding correct?

**Comment by angelblue05** [comment]
@dimitry-ishenko 
When setting StartPercent to 0, it was being ignored and Kodi went ahead and loaded the resume point instead of starting from the start.

**Comment by dimitry-ishenko** [comment]
@angelblue05 could you please tell me how to reproduce the bug? I'd like to try to revert the commit and see what's going on

**Comment by angelblue05** [comment]
Create a listitem with media type and db id of a video item in database that has a resume point. Set the listitem property StartPercent to 0. When you play with setresolvedurl, It'll load the existing resume point with this code here, because im_iStartOffset is -1 without my PR line:

https://github.com/angelblue05/xbmc/blob/f4f8c65bd371ee35e8b93c96878d6a7aca69bdfc/xbmc/Application.cpp#L2833-L2842

And here, startpercent would be completely ignored because startpercent value is 0 and starttime is filled with the resume point:
https://github.com/xbmc/xbmc/blob/8826ce905e4a05036aff3b7d6e5e2b8820fade15/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L1300-L1310

StartPercent is the only property that allows to control where playback start. I used to use "StartOffset" refered here but it was broken in Leia or never really supported even though it was in the docs: https://codedocs.xyz/xbmc/xbmc/group__python__xbmcgui__listitem.html#gaf3890e39353d2e8f46928653667e7d81

But in my opinion, my PR is the right thing to do, as the StartPercent is the resume value and Kodi should not be overwriting that. However, that is the solution we came with at the time:

https://github.com/xbmc/xbmc/blob/8826ce905e4a05036aff3b7d6e5e2b8820fade15/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L1300-L1302

**Comment by dimitry-ishenko** [comment]
I still wrapping my head around it, but my immediate reaction is that `startpercent` (and probably `starttime`) member for `CPlayerOptions` should be of type `std::optional<double>` to differentiate between "0" meaning start from 0% and "0" meaning `startpercent` is not set.

Unfortunately, `std::optional` is part of c++17 and I believe kodi is using c++11. So maybe we can use "-1" to indicate that `startpercent` is not set and "0" to start from 0%.

Not sure if this will break some other code. I will try to look into this in the next few days. If any of the devs familiar with this code could chime in, I would appreciate that as well.

**Comment by dimitry-ishenko** [comment]
@DaveTBlake thanks. That makes sense. I can try to tackle this, but it touches on quite a few different parts of code, so I might need a lot of hand-holding.

Maybe in the meantime we add a simple check to @angelblue05's patch:
`if (!item.IsAudioBook()) item.m_lStartOffset = 0;`

**Comment by angelblue05** [comment]
I'm not setup to compile Kodi atm. But I'm sure it won't interfere as there's no type audiobook for listitems. 

https://codedocs.xyz/xbmc/xbmc/group__python__xbmcgui__listitem.html#ga0b71166869bda87ad744942888fb5f14

**Comment by dimitry-ishenko** [comment]
OK with the above changes everything should be working now. One last thing I want to do is add a context menu similar to the one for videos with "Resume from ..." and "Play from beginning" items.

I would appreciate any input on what I've done so far.

**Review by notspiff — COMMENTED** [review]
I have not tested but it looks to be all good. In particular killing the sleep stuff is a welcome improvement

**Comment by MartijnKaijser** [comment]
Most is audiobook so I don't object. However I do want some approved marks from several devs

File: xbmc/cores/paplayer/PAPlayer.cpp [hunk 1]
```diff
@@ -373,7 +369,10 @@ bool PAPlayer::QueueNextFileEx(const CFileItem &file, bool fadeIn)
   si->m_finishing = false;
   si->m_framesSent = 0;
   si->m_seekNextAtFrame = 0;
-  si->m_seekFrame = -1;
+  if (si->m_fileItem.HasProperty("audiobook_bookmark"))
```
Comment (pkerling) [inline xbmc/cores/paplayer/PAPlayer.cpp]: I don't really like that the property is called `audiobook_bookmark` even though it just tells PAPlayer an offset to start playing at. This is not audiobook-specific, is it?

**Comment by dimitry-ishenko** [comment]
This should be it. It works pretty nicely for me. Any reviews and/or testing are welcome :smile:

**Comment by dimitry-ishenko** [comment]
@MartijnKaijser is anything else needed from me or are we just waiting for someone to find time to review?

**Comment by MartijnKaijser** [comment]
Thx
Feel free to continue on improving anything you might find

File: xbmc/cores/paplayer/PAPlayer.cpp (hunk 1 above)
Comment (pkerling) [reply xbmc/cores/paplayer/PAPlayer.cpp]: What about `audio_starttime` for example?

**Comment by HitcherUK** [comment]
@dimitry-ishenko sorry for asking here but I can't find any information on how audio books should be tagged so they're added to Kodi correctly. Thanks.

**Comment by dimitry-ishenko** [comment]
> @dimitry-ishenko sorry for asking here but I can't find any information on how audio books should be tagged so they're added to Kodi correctly. Thanks.

@HitcherUK all files with .mka and .m4b extensions are considered audio books:

https://github.com/xbmc/xbmc/blob/2f43cc1bbde193869ee11c690d14de82fc6913eb/xbmc/FileItem.cpp#L908-L911

**Comment by HitcherUK** [comment]
Thanks, I'd converted my .aax files to .mp3 so that was the problem with them not being recognised as audio books.

**Pull request merged**